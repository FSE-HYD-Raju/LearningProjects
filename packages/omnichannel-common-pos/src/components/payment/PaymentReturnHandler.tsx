import { isEqual, get, omit } from "lodash";
import * as React from "react";
import { Component } from "react";
import OcLoadingSpinner from "../ocComponents/OcLoadingSpinner";
import isClient from "../../utils/isClient";
import { RouteComponentProps } from "react-router";
import {
	Basket,
	BasketActionAddProductToBasket,
	BasketActionSubmitBasket,
	PaymentActionValidatePayment,
	ValidatedPaymentMethod,
	BasketActionAddProductToBasketOptions
} from "../../redux/types";

interface PaymentReturnHandlerStateProps {
	submittedBasket?: Basket;
	committedBasket?: Basket;
	paymentStatus?: string;
	errors: Array<any>;
	customerAccountId?: string;
	contextualPaymentMethodId: string;
	successRoute: string;
	rejectRoute: string;
	receiptCreateFailedRoute: string;
	disableBasketSubmit: boolean;
	storedTopupProduct?: BasketActionAddProductToBasketOptions;
	selectedPaymentUseCase: string;
	returnBaseLocation: Location;
	customerPaymentMethod: ValidatedPaymentMethod;
	paymentMethodRelationId?: string;
}

type PaymentReturnHandlerStateWithRouteProps = PaymentReturnHandlerStateProps & RouteComponentProps<any>;

interface PaymentReturnHandlerActionProps {
	actions: {
		submitBasket: BasketActionSubmitBasket;
		addProductToBasket: BasketActionAddProductToBasket;
		handlePaymentReject: () => void;
		validatePayment: PaymentActionValidatePayment;
		validatePaymentWithoutBasket: (
			paymentUsecase: string,
			contextualPaymentMethodId: string,
			paymentParams: object,
			customerAccountId: string
		) => void;
		historyPush: (path: string) => void;
	};
}

interface PaymentReturnHandlerState {
	loading: boolean;
	validating: boolean;
	submitting: boolean;
}

interface SearchParams {
	[key: string]: string;
}

type PaymentReturnHandlerProps = PaymentReturnHandlerStateWithRouteProps & PaymentReturnHandlerActionProps;

class PaymentReturnHandler extends Component<PaymentReturnHandlerProps, PaymentReturnHandlerState> {
	static displayName = "PaymentReturnHandler";
	static DEFAULT_KEY_PAYMENT_METHOD_RELATION = "T_Payment_Method_Relation";

	constructor(props: PaymentReturnHandlerProps) {
		super(props);
		this.state = {
			validating: false,
			loading: true,
			submitting: false
		};
	}

	componentDidMount() {
		if (isClient) {
			this.checkReceiptAndValidate(this.props);
		}
	}

	componentWillReceiveProps(nextProps: PaymentReturnHandlerProps) {
		if (
			isClient &&
			!isEqual(
				omit(this.props, ["actions", "flux", "history", "location", "match", "staticContext"]),
				omit(nextProps, ["actions", "flux", "history", "location", "match", "staticContext"])
			)
		) {
			this.checkReceiptAndValidate(nextProps);
		}
	}

	addPaymentMethodInformation = (props: PaymentReturnHandlerProps) => {
		const key = props.paymentMethodRelationId || PaymentReturnHandler.DEFAULT_KEY_PAYMENT_METHOD_RELATION;
		const topup = props.storedTopupProduct!;

		if (topup.product) {
			if (!topup.product.inputCharacteristics) {
				topup.product.inputtedCharacteristics = {};
			}
			if (topup.product.inputtedCharacteristics) {
				topup.product.inputtedCharacteristics[key] = get(props, "customerPaymentMethod.id", "");
			}
		}

		return topup;
	};

	// TODO: check this carefully
	submit = async (props: PaymentReturnHandlerProps, basket: Basket) => {
		if (!props.disableBasketSubmit) {
			if (!props.storedTopupProduct) {
				this.setState({ submitting: true },
					() => {
						props.actions.submitBasket(basket && basket.id)
							.then(() => {
								this.setState({ submitting: false });
							})
							.catch(() => {
								this.setState({ submitting: false });
							})
						;
					}
				);

			} else if (props.customerPaymentMethod) {
				const product = this.addPaymentMethodInformation(props);

				this.setState({ submitting: true },
					() => {
						props.actions.addProductToBasket(product)
							.then(() => {
								props.actions.submitBasket(basket && basket.id)
									.then(() => {
										this.setState({ submitting: false });
									})
								;
							})
							.catch(() => {
								this.setState({ submitting: false });
							})
						;
					}
				);
			}
		} else if (props.storedTopupProduct && props.customerPaymentMethod) {
			const product = this.addPaymentMethodInformation(props);

			this.setState({ submitting: true },
				() => {
					props.actions.addProductToBasket(product)
						.then(() => {
							this.setState({ submitting: false });
						})
						.catch(() => {
							this.setState({ submitting: false });
						})
					;
				}
			);
		}
	};

	parseSearchParams(search: string): SearchParams {
		const searchParams: SearchParams = {};
		search
			.trim()
			.substr(1)
			.split("&")
			.map((param: string) => {
				const kv = param.split("=");
				searchParams[kv[0]] = kv[1];
			});
		return searchParams;
	}

	checkReceiptAndValidate(props: PaymentReturnHandlerProps) {
		const {
			submittedBasket,
			committedBasket,
			customerAccountId,
			paymentStatus,
			errors,
			contextualPaymentMethodId,
			selectedPaymentUseCase,
			returnBaseLocation,
		} = props;

		if (committedBasket && !submittedBasket && !errors && !this.state.validating) {

			this.setState({ validating: true });
			const basketId = committedBasket.id;

			const paymentParams: SearchParams = this.parseSearchParams(this.props.location.search);

			if (customerAccountId) { // validation is only available for a real customer account
				props.actions.validatePayment(basketId, contextualPaymentMethodId, paymentParams, customerAccountId);
			}
		} else if (!committedBasket && !submittedBasket && !errors && !this.state.validating) {
			this.setState({ validating: true });
			const paymentParams: SearchParams = this.parseSearchParams(this.props.location.search);

			this.props.actions.validatePaymentWithoutBasket(
				selectedPaymentUseCase,
				contextualPaymentMethodId,
				paymentParams,
				""
			);
		} else if (paymentStatus) {
			if (paymentStatus === "SUCCESS" && submittedBasket && !returnBaseLocation) {
				this.props.actions.historyPush(this.props.successRoute);
			} else if (paymentStatus === "SUCCESS" && !submittedBasket && committedBasket && !this.state.submitting) {
				this.submit(props, committedBasket);
			} else if (paymentStatus === "SUCCESS" && returnBaseLocation && returnBaseLocation.pathname) {
				this.props.actions.historyPush(returnBaseLocation.pathname);
			} else if (paymentStatus === "REJECT") {
				this.props.actions.handlePaymentReject();
				this.props.actions.historyPush(this.props.rejectRoute);
			} else if (paymentStatus === "RECEIPT_CREATE_FAILED") {
				this.props.actions.historyPush(this.props.receiptCreateFailedRoute);
			}
		}
	}

	render() {
		return <OcLoadingSpinner loading={this.state.loading} />;
	}
}

export default PaymentReturnHandler;
export {
	PaymentReturnHandlerProps,
	PaymentReturnHandlerState,
	PaymentReturnHandlerStateProps,
	PaymentReturnHandlerStateWithRouteProps,
	PaymentReturnHandlerActionProps,
	SearchParams
};
