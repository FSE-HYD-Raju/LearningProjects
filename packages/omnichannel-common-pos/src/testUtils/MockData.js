import _ from "lodash";
import R from "ramda";

const firstDateThisWeek = () => {
	const date = new Date();
	while (date.getDay() !== 0) {
		date.setDate(date.getDate() - 1);
	}
	return date;
};

const people = [
	{
		firstName: "Kathrin",
		lastName: "K",
		gender: "female",
		language: "swe",
		countryOfBirth: "SE",
		birthDay: firstDateThisWeek(),
		emails: [{ email: "kathrin.k@qvantel.com" }],
		email: "kathrin.k@qvantel.com",
		mobileNumbers: [{ number: "2398457" }],
		mobileNumber: "2398457",
		fixedLineNumbers: [{ number: "2398457" }],
		fixedLineNumber: "2398457",
		password: "qwerasdf",
		passwordConfirmation: "qwerasdf"
	},
	{
		firstName: "John",
		lastName: "Smith",
		gender: "male",
		language: "fin",
		birthDay: firstDateThisWeek(),
		countryOfBirth: "United Kingdom",
		emails: [{ email: "john.smith@qvantel.com" }],
		email: "john.smith@qvantel.com",
		mobileNumbers: [{ number: "3498573" }],
		mobileNumber: "3498573",
		fixedLineNumbers: [{ number: "52497584" }],
		fixedLineNumber: "52497584"
	}
];

const basketItems = [
	{
		id: "7879c247-9e98-40ee-bb48-544eb366494a",
		type: "basketItems",
		attributes: {
			product: {
				name: "Movie Steaming Service"
			},
			totalPrices: [
				{
					type: "ONE_TIME",
					taxFreeAmount: 1.9,
					taxRate: 0,
					currency: "EUR"
				},
				{
					type: "RECURRENT",
					taxFreeAmount: 14.9,
					taxRate: 0,
					currency: "EUR"
				}
			],
			childBasketItems: [],
			totalUpfrontPrice: 1.9,
			totalUpfrontPriceWithChildren: 1.9
		}
	},
	{
		id: "c0065ffc-77f4-4e44-b15c-fbd2f78d04ce",
		type: "basketItems",
		attributes: {
			product: {
				name: "XDSL"
			},
			totalPrices: [
				{
					type: "ONE_TIME",
					taxFreeAmount: 100.6,
					taxRate: 0,
					currency: "EUR"
				},
				{
					type: "RECURRENT",
					taxFreeAmount: 30,
					taxRate: 0,
					currency: "EUR"
				}
			],
			childBasketItems: []
		}
	}
];

const deliveryItem = {
	id: "c0065ffc-77f4-4e44-b15c-fbd2f78d04ce",
	type: "basketItems",
	inputtedCharacteristics: {
		CH_Delivery_Contact_Info: "5c5dddf6-72be-4d0a-9711-6027f247af28"
	},
	product: {
		name: "Home delivery",
		id: "PO_Home_Delivery"
	}
};

export default class MockData {
	static getPerson(searchCriteria) {
		if (searchCriteria) {
			return _.find(people, searchCriteria);
		} else {
			return people[0];
		}
	}

	static getPeople() {
		return people;
	}

	static getPostalAddress() {
		return {
			street: "1 Test St",
			coAddress: "Best Ltd.",
			postalCode: "90210",
			city: "Test City",
			country: "FI"
		};
	}

	static getUser() {
		return {
			"id": "bf588945-ce95-490a-a774-8ef91661a146",
			"type": "persons",
			"attributes": {
				"lastName": "Testeri",
				"addresses": null,
				"gender": "male",
				"fixedLineNumbers": [],
				"formattedName": "Teppo Testeri",
				"roles": null,
				"billingAccountId": null,
				"language": null,
				"emails": [
					{
						"role": "PRIMARY",
						"id": "2ab79d7b-862d-49df-8afc-e6104a914919",
						"email": "tepsteri@tepinposti.fi"
					}
				],
				"honorificSuffix": null,
				"countryOfBirth": "FI",
				"personRelationships": [],
				"birthDay": null,
				"placeOfBirth": null,
				"characteristics": null,
				"mobileNumbers": [
					{
						"role": "PRIMARY",
						"id": "a843f46d-f729-40dd-a7ce-240a64bc008e",
						"number": "1234576890"
					}
				],
				"honorificPrefix": null,
				"avatar": "/img/icon-user-male.svg",
				"privacySettings": {
					"privacy1": false,
					"privacy2": false
				},
				"postalAddresses": [
					{
						"role": "PRIMARY",
						"id": "5e30221d-ad19-4043-8088-f594f5fdf77f",
						"street": "Piippukatu 11 A 1",
						"coAddress": null,
						"postalCode": "40100",
						"city": "Roma",
						"country": "FI",
						"county": null,
						"description": null,
						"postOfficeBox": null,
						"stateOrProvince": null,
						"addressRegisterId": null,
						"apartment": null,
						"building": null
					},
					{
						"role": "DELIVERY",
						"id": "5c5dddf6-72be-4d0a-9711-6027f247af28",
						"street": "Piippukatu 11 A 1",
						"coAddress": "Piippukatu 11 A 1",
						"postalCode": "40100",
						"city": "Roma",
						"country": "FI",
						"county": null,
						"description": null,
						"postOfficeBox": null,
						"stateOrProvince": "",
						"addressRegisterId": null,
						"apartment": null,
						"building": null
					}
				],
				"identifications": [],
				"firstName": "Teppo",
				"customerAccountId": null,
				"nationality": null,
				"additionalName": null,
				"maritalStatus": null
			}
		}
	}

	static getBasket(withItems: boolean, withDeliveryItems: boolean) {
		const basket = {
			activeBasket: {
				attributes: {
					totalPrices: [],
					totalUpfrontPrice: 0
				}
			}
		};

		if (withItems) {
			basket.basketItems = basketItems;

			const totalPriceOneTime = R.reduce(
				R.add,
				0,
				basketItems.map(bi => {
					return bi.attributes.totalPrices[0].taxFreeAmount;
				})
			);
			const totalPriceMonthly = R.reduce(
				R.add,
				0,
				basketItems.map(bi => {
					return bi.attributes.totalPrices[0].taxFreeAmount;
				})
			);

			basket.activeBasket.attributes.totalPrices = [
				{
					type: "ONE_TIME",
					taxFreeAmount: totalPriceOneTime,
					taxRate: 0,
					currency: "EUR"
				},
				{
					type: "RECURRENT",
					taxFreeAmount: totalPriceMonthly,
					taxRate: 0,
					currency: "EUR"
				}
			];

			if (withDeliveryItems) {
				basket.basketItems[0].attributes.childBasketItems.push(
					deliveryItem
				);
			}
		}

		return basket;
	}
}
