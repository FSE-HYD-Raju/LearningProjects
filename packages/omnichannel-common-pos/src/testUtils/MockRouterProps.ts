import {
	RouteComponentProps,

} from "react-router";

import {
	Href,
	UnregisterCallback,
	TransitionPromptHook,
	LocationState,
	Path,
	LocationDescriptorObject,
	LocationListener
} from "history";

const mockLocation: RouteComponentProps<any>["location"] = {
		hash: "",
		key: "",
		pathname: "",
		search: "",
		state: {}
};

const mockHistory: RouteComponentProps<any>["history"] = {
	length: 2,
	action: "POP",
	location: mockLocation,
	push: (path: Path | LocationDescriptorObject, state?: LocationState): void => {},
	replace: (path: Path | LocationDescriptorObject, state?: LocationState): void => {},
	go: (n: number): void => {},
	goBack: (): void => {},
	goForward: (): void => {},
	block: (prompt?: boolean | string | TransitionPromptHook): UnregisterCallback => { return () => {}; },
	listen: (listener: LocationListener): UnregisterCallback => {return () => {}; },
	createHref: (location: LocationDescriptorObject): Href => { return ""; },
};

const mockRouterProps: RouteComponentProps<any> = {
		match: {
			isExact: true,
			params: {},
			path: "",
			url: ""
		},
		location: mockLocation,
		history: mockHistory,
		staticContext: {}
};

export {
	mockRouterProps,
	mockLocation,
	mockHistory,
	RouteComponentProps
};
