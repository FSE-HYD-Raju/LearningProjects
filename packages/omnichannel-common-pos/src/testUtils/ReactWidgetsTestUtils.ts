import { get, toArray } from "lodash";
import { ReactWrapper } from "enzyme";

export default class ReactWidgetsTestUtils {
	static openDateTimePicker(dateTimePicker: ReactWrapper) {
		dateTimePicker.find(".rw-select button").simulate("click");
	}

	static getValidElementInOpenedDateTimePicker(dateTimePicker: ReactWrapper, idx: number) {
		const datesOnRange = dateTimePicker
			.find(".rw-calendar-popup table td[role='gridcell']")
			.filterWhere(node => {
				return node.find(".rw-off-range").length === 0;
			});

		return datesOnRange.at(idx);
	}

	static selectValidElementInOpenedDateTimePicker(dateTimePicker: ReactWrapper, idx: number) {
		ReactWidgetsTestUtils.getValidElementInOpenedDateTimePicker(dateTimePicker, idx).simulate("click");
	}

	static selectFirstValidElementInOpenedDateTimePicker(dateTimePicker: ReactWrapper) {
		ReactWidgetsTestUtils.selectValidElementInOpenedDateTimePicker(
			dateTimePicker,
			0
		);
	}

	static selectLastValidElementInOpenedDateTimePicker(dateTimePicker: ReactWrapper) {
		const datesOnRange = dateTimePicker
			.find(".rw-calendar-popup table td[role='gridcell']")
			.filterWhere(node => {
				return node.find(".rw-off-range").length === 0;
			});

		return datesOnRange.at(datesOnRange.getElements().length - 1);
	}

	static goForward(dateTimePicker: ReactWrapper) {
		dateTimePicker.find(".rw-btn-right").simulate("click");
	}

	static selectDate(dateTimePicker: ReactWrapper, date: Date) {
		const reId = new RegExp("^.*month_([0-1]?" + date.getMonth() + ")-([0-3]?" + date.getDate() + ")$");

		const dateCell = dateTimePicker.find("td").filterWhere((n: ReactWrapper) => {
			const nodeId = n.prop("id");
			if (!nodeId) {
				return false;
			}

			const idMatch = (n.prop("id") as any).match(reId);

			if (idMatch) {
				const [rMonthIdx, rDateIdx] = toArray(idMatch);
				return (
					parseInt(rMonthIdx, 10) === date.getMonth() && parseInt(rDateIdx, 10) === date.getDate()
				);
			} else {
				return false;
			}
		});

		try {
			dateCell.find("span").at(0).simulate("click");
		} catch (e) {
			if (e.message.match(/Method “simulate” is only meant to be run on a single node/) !== null) {
				throw new Error("Seeked date cell not found");
			} else {
				throw e;
			}
		}
	}
}