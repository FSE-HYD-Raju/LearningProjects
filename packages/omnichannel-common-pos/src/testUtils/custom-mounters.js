import React from "react";
import { IntlProvider, intlShape } from "react-intl";
import { mount, shallow } from "enzyme";
import PropTypes from "prop-types";
import R from "ramda";
import TestUtils from "./TestUtils";

const MAX_DIVE_DEEP_LEVEL = 16;
const intlProvider = new IntlProvider({ locale: "en" }, {});
const { intl } = intlProvider.getChildContext();

const baseContext = {
	flux: {
		persist: () => {}
	},
	store: TestUtils.mockReduxStore(),
	intl,
	location: {},
	only: {},
	skip: {}
};

const nodeWithIntlProp = (node, mockedRouterOptions) => {
	if (mockedRouterOptions) {
		mockedRouterOptions.props().history.push = jest.fn();
	}
	return React.cloneElement(node, {
		intl,
		...(mockedRouterOptions ? mockedRouterOptions.props() : {})
	});
};

function getOptionsForShallowMount(options, debug, mockedRouterContextOptions) {
	const mountOptions = R.merge(options ? options : {}, {
		disableLifecycleMethods: !(
			options && options.disableLifecycleMethods === false
		),
		context: R.mergeAll([
			debug
				? {
						...baseContext
					}
				: baseContext,
			R.omit("intl", options && options.context ? options.context : {}),
			mockedRouterContextOptions.context
		]),
		childContextTypes: R.merge(
			{
				intl: intlShape,
				store: PropTypes.object
			},
			mockedRouterContextOptions.childContextTypes
		)
	});
	TestUtils.mockReduxStoreToFlux(mountOptions.context);
	return mountOptions;
}
function deepDive(wrapper) {
	let diveAttemptsCount = 0;
	while (
		wrapper.hostNodes().length === 0 &&
		diveAttemptsCount < MAX_DIVE_DEEP_LEVEL
	) {
		wrapper = wrapper.dive();
		diveAttemptsCount += 1;
	}
	return wrapper;
}
export function shallowWithContextAndRouterProps(node, options, debug) {
	return shallowWithContext(node, options, debug, true);
}
export function shallowWithContextAndLifecycle(node, options, debug) {
	return shallowWithContext(
		node,
		R.merge(options || {}, { disableLifecycleMethods: false }),
		debug,
		true
	);
}
export function deepShallowWithContextAndLifecycle(node, options, debug) {
	return deepDive(shallowWithContextAndLifecycle(node, options, debug));
}
export function deepShallowWithContext(node, options, debug) {
	return deepDive(shallowWithContext(node, options, debug, true));
}
export function shallowWithContext(node, options, debug, injectRouterProps) {
	const mockedRouterOptions = TestUtils.getMockedRouterOptions();
	const mockedRouterContextOptions = mockedRouterOptions.get();
	return shallow(
		nodeWithIntlProp(node, injectRouterProps ? mockedRouterOptions : null),
		getOptionsForShallowMount(options, debug, mockedRouterContextOptions)
	);
}

export function mountWithContextAndRouterProps(node, options, debug) {
	return mountWithContext(node, options, debug, true);
}
export function mountWithContext(node, options, debug, injectRouterProps) {
	const mockedRouterOptions = TestUtils.getMockedRouterOptions();
	const mockedRouterContextOptions = mockedRouterOptions.get();
	const mountOptions = R.merge(options ? options : {}, {
		context: R.mergeAll([
			debug
				? {
						...baseContext
					}
				: baseContext,
			R.omit("intl", options && options.context ? options.context : {}),
			mockedRouterContextOptions.context
		]),
		childContextTypes: R.merge(
			{
				flux: PropTypes.object,
				store: PropTypes.object,
				intl: intlShape,
				location: PropTypes.object,
				only: PropTypes.object,
				skip: PropTypes.object
			},
			mockedRouterContextOptions.childContextTypes
		)
	});
	TestUtils.mockReduxStoreToFlux(mountOptions.context);
	return mount(
		nodeWithIntlProp(node, injectRouterProps ? mockedRouterOptions : null),
		mountOptions
	);
}

const attachmentElementId = "__OC_COMPONENT_TEST_ATTACHMENT_ELEMENT__";

export function attachWithContext(node, options) {
	const attachmentElement = document.createElement("div");
	attachmentElement.setAttribute("id", attachmentElementId);
	document.body.appendChild(attachmentElement);

	const wrapper = mountWithContext(
		node,
		R.merge(options, { attachTo: attachmentElement })
	);
	const doppelWrapper = hijackDetacher(wrapper);

	return doppelWrapper;
}

function hijackDetacher(wrapper) {
	const doppel = wrapper;
	doppel.__detach = wrapper.detach;

	doppel.detach = () => {
		removeAttachmentElements();
		doppel.__detach.apply(wrapper);
	};

	return doppel;
}

function removeAttachmentElements() {
	document.body.querySelectorAll("#" + attachmentElementId).forEach(elem => {
		document.body.removeChild(elem);
	});
}

export function detach(wrapper) {
	removeAttachmentElements();
	wrapper.detach();
}
