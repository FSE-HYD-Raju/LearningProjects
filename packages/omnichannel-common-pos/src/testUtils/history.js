import isClient from "../utils/isClient";
import { createBrowserHistory, createMemoryHistory } from "history";
import { useRouterHistory } from "react-router";

const history = isClient
	? useRouterHistory(createBrowserHistory)({ queryKey: false })
	: createMemoryHistory();

export default history;
