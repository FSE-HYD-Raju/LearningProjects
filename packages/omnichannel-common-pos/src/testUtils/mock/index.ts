import { default as productOffering } from "./ProductOfferingMock";
import { ProductMock as product } from "./ProductMock";
import { AgreementMock as agreement } from "./AgreementMock";
import { ChargingBalancesMock as chargingBalances } from "./ChargingBalancesMock";
import { default as postalAddress } from "./PostalAddressMock";
import { default as identification } from "./IdentificationMock";
import { default as user } from "./UserMock";
import { default as order } from "./OrderMock";
import { default as basket } from "./BasketMock";
export { productOffering, postalAddress, identification, user, order, basket, product, agreement, chargingBalances };
