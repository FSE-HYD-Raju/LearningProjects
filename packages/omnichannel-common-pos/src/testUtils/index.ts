import ReactWidgetsTestUtils from "./ReactWidgetsTestUtils";
import * as MockDataMaker from "./mock";
const {
	shallowWithContextAndRouterProps,
	shallowWithContextAndLifecycle,
	deepShallowWithContextAndLifecycle,
	deepShallowWithContext,
	shallowWithContext,
	mountWithContextAndRouterProps,
	mountWithContext,
	attachWithContext,
	detach
} = require("./custom-mounters");
const TestUtils = require("./TestUtils");
const MockData = require("./MockData");
const SimpleDataMock = require("./mock/simpleDataMock");

export {
	MockData,
	ReactWidgetsTestUtils,
	TestUtils,
	SimpleDataMock,
	shallowWithContextAndRouterProps,
	shallowWithContextAndLifecycle,
	deepShallowWithContextAndLifecycle,
	deepShallowWithContext,
	shallowWithContext,
	mountWithContextAndRouterProps,
	mountWithContext,
	attachWithContext,
	detach,
	MockDataMaker
};

export * from "./MockRouterProps";

export default TestUtils;
