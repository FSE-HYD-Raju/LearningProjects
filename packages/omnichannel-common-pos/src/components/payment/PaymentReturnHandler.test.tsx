import * as React from "react";
import { mountWithContext, shallowWithContext, TestUtils } from "../../testUtils";

import PaymentReturnHandler, { PaymentReturnHandlerStateProps, PaymentReturnHandlerActionProps } from "./PaymentReturnHandler";

describe("PaymentReturnHandler", () => {
	const creditCard = {
		type: "credit-card",
		id: "masas-credit-card",
		name: "Masa's Credit Card",
		cardHolder: "Masa Meikäläinen",
		cardType: "digital-mint",
		expiryMonth: "3",
		expiryYear: "2022",
		maskedCardNumber: "1234",
		token: "all-your-monneh-are-belong-to-masa",
	};

	const customerPaymentMethod = {
		id: "payment-with-credit-card",
		name: "Best Payment with Credit",
		type: "Credit payment",
		creditCard,
		validFor: {
			startDate: new Date(2019, 2, 12, 0, 0, 0, 0),
			endDate: new Date(2022, 2, 12, 0, 0, 0, 0)
		},
		role: "PRIMARY"
	};

	const topUpProduct = {
		product: {},
		hasCustomer: true
	};

	const paymentMethodRelationId = "payment-method-relation-id";

	let minimumProps: any;
	beforeEach(() => {
		minimumProps = {
			errors: [],
			contextualPaymentMethodId: "brilliant-contextual-payment-method-id",
			successRoute: "great-payment-success",
			rejectRoute: "cruel-rejection-page",
			receiptCreateFailedRoute: "receipt-creation-failed",
			disableBasketSubmit: false,
			storedTopupProduct: undefined,
			selectedPaymentUseCase: "selected-payment-use-case",
			returnBaseLocation: {} as any,
			customerPaymentMethod: {} as any,
			actions: {
				submitBasket: (() => {}) as any,
				addProductToBasket: (() => {}) as any,
				handlePaymentReject: () => {},
				validatePayment: () => {},
				validatePaymentWithoutBasket: () => {},
				historyPush: () => {},
			}
		} as PaymentReturnHandlerStateProps & PaymentReturnHandlerActionProps;
	});

	it("succeeds at shallow mount with minimum props", () => {
		const wrapper = shallowWithContext(
			<PaymentReturnHandler {...minimumProps} />
		);
		expect(wrapper).toMatchSnapshot();
	});

	it("succeeds at deep mount with minimum props", () => {
		mountWithContext(
			<PaymentReturnHandler {...minimumProps} />
		);
	});

	it("submits basket at mount, no top-up product stored", (done) => {
		const props = {
			...minimumProps,
			actions: {
				...minimumProps.actions,
				submitBasket: (basket: any) => {
					return new Promise(() => {
						done();
					});
				}
			},
			paymentStatus: "SUCCESS",
			committedBasket: {
				id: "foobar-committed-basket-id"
			}
		};

		mountWithContext(<PaymentReturnHandler {...props} />);
	});

	it("submits basket for a top-up product", (done) => {
		const initialProps = {
			...minimumProps,
			actions: {
				...minimumProps.actions,
				addProductToBasket: (product: any) => {
					return new Promise((resolve, reject) => {
						resolve();
					});
				},
				submitBasket: (basket: any) => {
					return new Promise(() => {
						done();
					});
				}
			},
			storedTopupProduct: topUpProduct,
			paymentStatus: "SUCCESS",
			committedBasket: {
				id: "foobar-committed-basket-id"
			}
		};

		let wrapper = mountWithContext(
			<PaymentReturnHandler {...initialProps} />
		);

		const nextProps = {
			...initialProps,
			foo: "bar"
		};

		wrapper = wrapper.setProps(nextProps);
	});

	it("adds payment information on top-up product", (done) => {
		const props = {
			...minimumProps,
			actions: {
				...minimumProps.actions,
				addProductToBasket: (product: any) => {
					expect(product).toEqual(
						{
							...topUpProduct,
							product: {
								inputtedCharacteristics: {
									[paymentMethodRelationId]: props.customerPaymentMethod.id
								}
							}
						}
					);

					return new Promise(() => {
						done();
					});
				}
			},
			disableBasketSubmit: true,
			storedTopupProduct: topUpProduct,
			paymentMethodRelationId,
			customerPaymentMethod,
			paymentStatus: "SUCCESS",
			committedBasket: {
				id: "foobar-committed-basket-id"
			}
		};

		mountWithContext(<PaymentReturnHandler {...props} />);
	});

	/* RUBT-143856
	 *
	 * 1. basket is submitted at mount
	 * 2. prop submittedBasket and submitting:true in state block sequential calls to actions.submitBasket().
	 */
	describe("basket submit during component lifetime", () => {
		let defaultProps: any;
		beforeEach(() => {
			defaultProps = {
				errors: [],
				contextualPaymentMethodId: "brilliant-contextual-payment-method-id",
				successRoute: "great-payment-success",
				rejectRoute: "cruel-rejection-page",
				receiptCreateFailedRoute: "receipt-creation-failed",
				disableBasketSubmit: false,
				selectedPaymentUseCase: "selected-payment-use-case",
				returnBaseLocation: {} as any,
				customerPaymentMethod: {} as any,
				storedTopupProduct: undefined,
				paymentStatus: "SUCCESS",
				committedBasket: {
					id: "foobar-committed-basket-id"
				} as any,
				actions: {
					submitBasket: (() => {}) as any,
					addProductToBasket: (() => {}) as any,
					handlePaymentReject: () => {},
					validatePayment: () => {},
					validatePaymentWithoutBasket: () => {},
					historyPush: () => {},
				}
			} as PaymentReturnHandlerStateProps & PaymentReturnHandlerActionProps;
		});

		it("submits basket at mount", (done) => {
			let callCounter = 0;

			const initialProps = {
				...defaultProps,
				actions: {
					...defaultProps.actions,
					submitBasket: (basket: any) => {
						++ callCounter;
						return new Promise(() => {});
					}
				}
			};

			let wrapper = mountWithContext(
				<PaymentReturnHandler {...initialProps} />
			);

			const nextProps = {
				...initialProps,
				foo: "bar"
			};

			wrapper = wrapper.setProps(nextProps);
			setTimeout(() => {
				expect(callCounter).toEqual(1);
				done();
			}, 500);
		});

		it("submits basket at mount, and second attempt with new props is blocked by state flag", (done) => {
			let callCounter = 0;
			const { sleep } = TestUtils;

			const secondProps = {
				...defaultProps
			};

			const initialProps = {
				...defaultProps,
				actions: {
					...defaultProps.actions,
					submitBasket: (basket: any) => {
						expect(basket).toBeTruthy();
						expect(basket).toEqual(defaultProps.committedBasket.id);

						return new Promise(async (resolve, reject) => {
							++ callCounter;

							await sleep(1000);
							resolve();

							wrapper = wrapper.setProps(secondProps);

							setTimeout(() => {
								expect(callCounter).toEqual(1);
								done();
							}, 500);
						});
					}
				},
			};

			secondProps.actions = initialProps.actions;
			secondProps.submittedBasket = {
				id: "foobar-submitted-basket-id"
			};

			let wrapper = mountWithContext(
				<PaymentReturnHandler {...initialProps} />
			);
		});
	});
});
