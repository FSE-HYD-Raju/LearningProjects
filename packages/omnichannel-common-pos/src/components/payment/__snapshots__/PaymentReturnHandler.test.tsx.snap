// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentReturnHandler succeeds at shallow mount with minimum props 1`] = `
<OcLoadingSpinner
  loading={true}
/>
`;
